File: gommo-frontend/src/modules/rh/person/jobvacancy/lib/job-vacancy.mapper.ts
import type { JobVacancy, JobVacancyCreateDto } from "@/modules/rh/person/jobvacancy/dto/job-vacancy.dto";
import { slugFromJobTitle } from "@/modules/rh/person/jobvacancy/lib/job-vacancy-slug";

function moneyToForm(value?: number | null): string {
    return value != null ? String(value) : "";
}

function moneyToPayload(value?: string | number): number | undefined {
    if (value == null || value === "") return undefined;
    return Number(value);
}

export function jobVacancyToFormDto(entity: JobVacancy): JobVacancyCreateDto {
    return {
        jobPositionId: entity.jobPositionId ?? null,
        jobTitle: entity.jobTitle ?? "",
        positionsCount: entity.positionsCount ?? 1,
        description: entity.description ?? "",
        activities: entity.activities ?? "",
        assignments: entity.assignments ?? "",
        requirements: entity.requirements ?? "",
        benefits: entity.benefits ?? "",
        department: entity.department ?? "",
        location: entity.location ?? "",
        workModality: entity.workModality,
        contractType: entity.contractType,
        workSchedule: entity.workSchedule ?? "",
        seniorityLevel: entity.seniorityLevel,
        salary: moneyToForm(entity.salary),
        salaryMax: moneyToForm(entity.salaryMax),
        expectedCompletionDate: entity.expectedCompletionDate?.slice(0, 10) ?? "",
        targetBoards: entity.targetBoards ?? [],
        slug: entity.slug ?? "",
        isPublic: Boolean(entity.isPublic),
    };
}

export const emptyJobVacancyForm = (): JobVacancyCreateDto => ({
    jobPositionId: null,
    jobTitle: "",
    positionsCount: 1,
    description: "",
    activities: "",
    assignments: "",
    requirements: "",
    benefits: "",
    department: "",
    location: "",
    workSchedule: "",
    salary: "",
    salaryMax: "",
    expectedCompletionDate: "",
    targetBoards: [],
    slug: "",
    isPublic: false,
});

export function jobVacancyFormToPayload(form: JobVacancyCreateDto): JobVacancyCreateDto {
    const isPublic = Boolean(form.isPublic);
    const slug =
        slugFromJobTitle(form.slug ?? "") ||
        (isPublic ? slugFromJobTitle(form.jobTitle) : "") ||
        undefined;
    return {
        ...form,
        jobPositionId: form.jobPositionId?.trim() || null,
        jobTitle: form.jobTitle.trim(),
        description: form.description?.trim() || undefined,
        activities: form.activities?.trim() || undefined,
        assignments: form.assignments?.trim() || undefined,
        requirements: form.requirements?.trim() || undefined,
        benefits: form.benefits?.trim() || undefined,
        department: form.department?.trim() || undefined,
        location: form.location?.trim() || undefined,
        workSchedule: form.workSchedule?.trim() || undefined,
        salary: moneyToPayload(form.salary),
        salaryMax: moneyToPayload(form.salaryMax),
        expectedCompletionDate: form.expectedCompletionDate?.trim() || undefined,
        targetBoards: form.targetBoards ?? [],
        slug,
        isPublic,
    };
}

File: gommo-backend/src/main/java/br/com/gommo/modules/rh/person/jobvacancy/mapper/JobVacancyMapper.java
package br.com.gommo.modules.rh.person.jobvacancy.mapper;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Component;

import br.com.gommo.modules.rh.person.jobvacancy.dto.JobVacancyRequestDto;
import br.com.gommo.modules.rh.person.jobvacancy.dto.JobVacancyResponseDto;
import br.com.gommo.modules.rh.person.jobvacancy.entity.JobVacancy;

@Component
public class JobVacancyMapper {
    public JobVacancy toEntity(JobVacancyRequestDto dto) {
        return JobVacancy.builder()
                .jobPositionId(dto.getJobPositionId())
                .jobTitle(requireTitle(dto.getJobTitle()))
                .positionsCount(dto.getPositionsCount())
                .description(trimToNull(dto.getDescription()))
                .activities(trimToNull(dto.getActivities()))
                .assignments(trimToNull(dto.getAssignments()))
                .requirements(trimToNull(dto.getRequirements()))
                .benefits(trimToNull(dto.getBenefits()))
                .department(trimToNull(dto.getDepartment()))
                .location(trimToNull(dto.getLocation()))
                .workModality(dto.getWorkModality())
                .contractType(dto.getContractType())
                .workSchedule(trimToNull(dto.getWorkSchedule()))
                .seniorityLevel(dto.getSeniorityLevel())
                .salary(dto.getSalary())
                .salaryMax(dto.getSalaryMax())
                .expectedCompletionDate(dto.getExpectedCompletionDate())
                .targetBoards(copyBoards(dto.getTargetBoards()))
                .slug(trimToNull(dto.getSlug()))
                .isPublic(Boolean.TRUE.equals(dto.getIsPublic()))
                .build();
    }

    public void updateEntity(JobVacancy entity, JobVacancyRequestDto dto) {
        entity.setJobPositionId(dto.getJobPositionId());
        entity.setJobTitle(requireTitle(dto.getJobTitle()));
        entity.setPositionsCount(dto.getPositionsCount());
        entity.setDescription(trimToNull(dto.getDescription()));
        entity.setActivities(trimToNull(dto.getActivities()));
        entity.setAssignments(trimToNull(dto.getAssignments()));
        entity.setRequirements(trimToNull(dto.getRequirements()));
        entity.setBenefits(trimToNull(dto.getBenefits()));
        entity.setDepartment(trimToNull(dto.getDepartment()));
        entity.setLocation(trimToNull(dto.getLocation()));
        entity.setWorkModality(dto.getWorkModality());
        entity.setContractType(dto.getContractType());
        entity.setWorkSchedule(trimToNull(dto.getWorkSchedule()));
        entity.setSeniorityLevel(dto.getSeniorityLevel());
        entity.setSalary(dto.getSalary());
        entity.setSalaryMax(dto.getSalaryMax());
        entity.setExpectedCompletionDate(dto.getExpectedCompletionDate());
        entity.setTargetBoards(copyBoards(dto.getTargetBoards()));
        entity.setSlug(trimToNull(dto.getSlug()));
        entity.setIsPublic(Boolean.TRUE.equals(dto.getIsPublic()));
    }

    public JobVacancyResponseDto toResponse(JobVacancy entity) {
        return toResponse(entity, 0);
    }

    public JobVacancyResponseDto toResponse(JobVacancy entity, int candidateCount) {
        return JobVacancyResponseDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .status(entity.getStatus())
                .jobPositionId(entity.getJobPositionId())
                .jobTitle(entity.getJobTitle())
                .positionsCount(entity.getPositionsCount())
                .description(entity.getDescription())
                .activities(entity.getActivities())
                .assignments(entity.getAssignments())
                .requirements(entity.getRequirements())
                .benefits(entity.getBenefits())
                .department(entity.getDepartment())
                .location(entity.getLocation())
                .workModality(entity.getWorkModality())
                .contractType(entity.getContractType())
                .workSchedule(entity.getWorkSchedule())
                .seniorityLevel(entity.getSeniorityLevel())
                .salary(entity.getSalary())
                .salaryMax(entity.getSalaryMax())
                .expectedCompletionDate(entity.getExpectedCompletionDate())
                .targetBoards(copyBoards(entity.getTargetBoards()))
                .slug(entity.getSlug())
                .isPublic(Boolean.TRUE.equals(entity.getIsPublic()))
                .publishedAt(entity.getPublishedAt())
                .candidateCount(candidateCount)
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .build();
    }

    private static List<String> copyBoards(List<String> boards) {
        if (boards == null || boards.isEmpty()) {
            return new ArrayList<>();
        }
        return new ArrayList<>(boards);
    }

    private static String requireTitle(String value) {
        return value == null ? "" : value.trim();
    }

    private static String trimToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}
